Dataset → student.csv

• To - Do - 1:
1. Read and Observe the Dataset.
2. Print top(5) and bottom(5) of the dataset {Hint: pd.head and pd.tail}.
3. Print the Information of Datasets. {Hint: pd.info}.
4. Gather the Descriptive info about the Dataset. {Hint: pd.describe}
5. Split your data into Feature (X) and Label (Y).

In [44]:
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/ConceptofAIandTechnology/student.csv")

#print top5 of the dataset
print("Top 5 data of dataset: ")
print(df.head())

#print bottom5 of dataset
print("\nBottom 5 data of dataset: ")
print(df.tail())

#print information of dataset
print("\nInformation of dataset:")
print(df.info)

#print descriptive info about the dataset
print("\nDescriptive info about the dataset:")
print(df.describe)

Top 5 data of dataset: 
   Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62

Bottom 5 data of dataset: 
     Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

Information of dataset:
<bound method DataFrame.info of      Math  Reading  Writing
0      48       68       63
1      62       81       72
2      79       80       78
3      76       83       79
4      59       64       62
..    ...      ...      ...
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

[1000 rows x 3 columns]>

Descriptive info about the dataset:
<bound method NDFrame.describe of      Math  Reading  Writing
0      48       68       63
1      62       81       72
2      79       80       78
3      76       83    

In [50]:
# Define the label column name
label_column = 'Writing'

# Separate features (X) and label (Y)
X = df.drop(columns=[label_column])  # Drop the label column to get features
Y = df[label_column]  # Extract the label column

print("\nFeatures (X):")
print(X.head())
print("\nLabel (Y):")
print(Y.head())


Features (X):
   Math  Reading
0    48       68
1    62       81
2    79       80
3    76       83
4    59       64

Label (Y):
0    63
1    72
2    78
3    79
4    62
Name: Writing, dtype: int64


• To - Do - 2:
1. To make the task easier - let’s assume there is no bias or intercept.
2. Create the following matrices:

Y = WTX

W =





w1
w2
.
.
.
wd





, where W ∈ R
d

X =





x1,1 x1,2 · · · x1,n
x2,1 x2,2 · · · x2,n
.
.
.
.
.
.
.
.
.
.
.
.
xd,1 xd,2 · · · xd,n





, where X ∈ R
d×n

Y =





y1
y2
.
.
.
yn





, where Y ∈ R
n

3. Note: The feature matrix described above does not include a column of 1s, as it assumes the
absence of a bias term in the model.

In [46]:
import numpy as np

# Define the dimensions
d = 3  # Number of features
n = 5  # Number of samples/examples

# Create W (weight vector) of shape (d, 1)
W = np.random.rand(d, 1)

# Create X (feature matrix) of shape (d, n)
X = np.random.rand(d, n)

# Compute Y = W^T X
Y = np.dot(W.T, X)

# Display matrices
print("Matrix W (weights):")
print(W)

print("\nMatrix X (features):")
print(X)

print("\nMatrix Y (output):")
print(Y)


Matrix W (weights):
[[0.27016076]
 [0.71022284]
 [0.5189397 ]]

Matrix X (features):
[[0.27858055 0.13618714 0.9070388  0.48704035 0.67271682]
 [0.15401911 0.1910896  0.63651863 0.48073921 0.7604193 ]
 [0.76656196 0.24572173 0.56569961 0.73170692 0.42175238]]

Matrix Y (output):
[[0.58244886 0.30002338 0.99068034 0.85272292 0.94067289]]


• To - Do - 3:
1. Split the dataset into training and test sets.
2. You can use an 80-20 or 70-30 split, with 80% (or 70%) of the data used for training and the rest
for testing.

In [51]:
from sklearn.model_selection import train_test_split

# Split the dataset into training and test sets using an 80-20 split
test_size = 0.2

# Split the dataset
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=test_size, random_state=42)

# Display the shapes of the splits
print("Training set:")
print(f"Features (X_train): {X_train.shape}, Labels (Y_train): {Y_train.shape}")

print("\nTest set:")
print(f"Features (X_test): {X_test.shape}, Labels (Y_test): {Y_test.shape}")


Training set:
Features (X_train): (800, 2), Labels (Y_train): (800,)

Test set:
Features (X_test): (200, 2), Labels (Y_test): (200,)


• To - Do:
We will define a function that:
1. Loads the data and splits it into training and test sets.
2. Prepares the feature matrix (X) and target vector (Y).
3. Defines the weight matrix (W) and initializes the learning rate and number of iterations.
4. Calls the gradient descent function to learn the parameters.
5. Evaluates the model using RMSE and R2
.

In [52]:
import pandas as pd
df = pd.read_csv("/content/drive/MyDrive/ConceptofAIandTechnology/student.csv")

# Define the label column name
label_column = 'Writing'

# Separate features (X) and label (Y)
X = df.drop(columns=[label_column])
Y = df[label_column]

print("\nFeatures (X):")
print(X.head())

print("\nLabel (Y):")
print(Y.head())




Features (X):
   Math  Reading
0    48       68
1    62       81
2    79       80
3    76       83
4    59       64

Label (Y):
0    63
1    72
2    78
3    79
4    62
Name: Writing, dtype: int64


In [53]:
from sklearn.model_selection import train_test_split

# Split the dataset into training and test sets using an 80-20 split
test_size = 0.2

# Perform the split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=test_size, random_state=42)

# Display the shapes of the splits
print(f"Training set shape (X_train, Y_train): {X_train.shape}, {Y_train.shape}")
print(f"Test set shape (X_test, Y_test): {X_test.shape}, {Y_test.shape}")


Training set shape (X_train, Y_train): (800, 2), (800,)
Test set shape (X_test, Y_test): (200, 2), (200,)


In [54]:
# Gradient Descent Function
def gradient_descent(X, Y, learning_rate=0.01, iterations=1000):
    m = len(Y)  # Number of samples
    d = X.shape[1]  # Number of features
    W = np.zeros(d)  # Initialize weight vector

    # Gradient Descent Loop
    for _ in range(iterations):
        # Compute predictions
        predictions = np.dot(X, W)

        # Calculate gradient
        gradient = - (2/m) * np.dot(X.T, (Y - predictions))

        # Update weights
        W -= learning_rate * gradient

    return W

# Train the model using gradient descent on the training set
W = gradient_descent(X_train.values, Y_train.values, learning_rate=0.01, iterations=1000)

# Display learned weights
print(f"Learned weights: {W}")


Learned weights: [nan nan]


<ipython-input-54-458b8ff31f41>:16: RuntimeWarning: invalid value encountered in subtract
  W -= learning_rate * gradient


In [56]:
from sklearn.metrics import mean_squared_error, r2_score

# Predict using the trained model
Y_pred = np.dot(X_test, W)

# If Y_pred contains NaN, replace NaNs with 0 (or any other appropriate value)
Y_pred = np.nan_to_num(Y_pred, nan=0.0)

# Calculate RMSE (Root Mean Squared Error)
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))

# Calculate R² (R-squared)
r2 = r2_score(Y_test, Y_pred)

# Display the evaluation metrics
print(f"RMSE: {rmse}")
print(f"R²: {r2}")


RMSE: 70.19622496972326
R²: -18.685008543136547


1. Did your Model Overfitt, Underfitts, or performance is acceptable.
-> The model is underfitting as indicated by the high RMSE (70.20) and negative R² (-18.69). This suggests that it is not capturing the underlying patterns in the data. To improve performance, try tuning the learning rate, increasing the number of iterations, or enhancing the features used for training.

2. Experiment with different value of learning rate, making it higher and lower, observe the result.

In [65]:
# Experiment with different learning rates
learning_rates = [0.001, 0.01, 0.1, 1, 10]
results = {}

for lr in learning_rates:
    # Train the model using gradient descent on the training set
    W = gradient_descent(X_train.values, Y_train.values, learning_rate=lr, iterations=1000)

    # Predict using the trained model
    Y_pred = np.dot(X_test, W)

    # If Y_pred contains NaN, replace NaNs with 0 (or any other appropriate value)
    Y_pred = np.nan_to_num(Y_pred, nan=0.0)

    # Calculate RMSE (Root Mean Squared Error)
    rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))

    # Calculate R² (R-squared)
    r2 = r2_score(Y_test, Y_pred)

    # Store the results
    results[lr] = {'RMSE': rmse, 'R²': r2}

# Display the results for each learning rate
for lr, result in results.items():
    print(f"Learning Rate: {lr}")
    print(f"RMSE: {result['RMSE']}")
    print(f"R²: {result['R²']}")
    print('-' * 40)


Learning Rate: 0.001
RMSE: 70.19622496972326
R²: -18.685008543136547
----------------------------------------
Learning Rate: 0.01
RMSE: 70.19622496972326
R²: -18.685008543136547
----------------------------------------
Learning Rate: 0.1
RMSE: 70.19622496972326
R²: -18.685008543136547
----------------------------------------
Learning Rate: 1
RMSE: 70.19622496972326
R²: -18.685008543136547
----------------------------------------
Learning Rate: 10
RMSE: 70.19622496972326
R²: -18.685008543136547
----------------------------------------


<ipython-input-54-458b8ff31f41>:16: RuntimeWarning: invalid value encountered in subtract
  W -= learning_rate * gradient
<ipython-input-54-458b8ff31f41>:16: RuntimeWarning: invalid value encountered in subtract
  W -= learning_rate * gradient
<ipython-input-54-458b8ff31f41>:16: RuntimeWarning: invalid value encountered in subtract
  W -= learning_rate * gradient
<ipython-input-54-458b8ff31f41>:16: RuntimeWarning: invalid value encountered in subtract
  W -= learning_rate * gradient
<ipython-input-54-458b8ff31f41>:16: RuntimeWarning: invalid value encountered in subtract
  W -= learning_rate * gradient
